### PR TITLE
MAR Model

### PR DESCRIPTION
**Pull request by inespancorbo**
Code is in model_code and preprocessing py scripts. The parameter (y - X.dot(betas[:, i-1])).T.dot(y - X.dot(betas[:, i-1])) / 2 in the inverse gamma (for sampling sigma2) increases with each iteration until it gets to infinity -- causing everything to crash. Left a notebook so you guys can see the behavior. Let me know what you think/have any ideas.

### CODE REVIEW

**Review by lbotti319 — COMMENTED** [review]
Can you re-run the notebook, so we can see the working output?

**Review by lbotti319 — COMMENTED** [review]
I think numeric values instead of histograms will be better for visualizing the difference between higher_yes and higher_no

File: preprocessing.py [hunk 1]
```diff
@@ -29,4 +29,45 @@ def prepare_data(df):
     encoded_cat_features = transform.get_feature_names(cat_cols)
     df_enc = pd.DataFrame.sparse.from_spmatrix(encoded_cats, columns = encoded_cat_features)
 
-    return X.join(df_enc), y
\ No newline at end of file
+    return X.join(df_enc), y
+
+
+def prepare_data_no_standardizing(df):
```
Comment (lbotti319) [inline preprocessing.py:35]: Want to just use the my preprocessing function instead? It also selects the features we are going to need.

File: model_code.py [hunk 2]
```diff
@@ -92,4 +91,139 @@ def Gibbs_regression(X, y, B):
     return betas[:,B:], accept[B:]
 
 
+###################################################################
+######### functions for missing data part of project ##############
+###################################################################
+
+
+def MH_step(param, log_target, log_proposal, accepts_param, *args):
+    param_star = log_proposal.rvs()
```
[the remark this answers is not in the document]
Comment (inespancorbo) [reply model_code.py:101]: Yea I don't know why I named it log_proposal, typo -- its using the proposal (normal). I ended not using this function though.

File: model_code.py [hunk 3]
```diff
@@ -92,4 +92,181 @@ def Gibbs_regression(X, y, B):
...
+    sigmas2[0] = sigma2
+
+    accepts_alpha0, accepts_alpha1, accepts_gamma0, accepts_gamma1 = 0, 0, 0, 0
+    ##################################################
+
+    ############ Sampling Gibbs + MH #################
+    proposal_alpha0 = norm(loc=0, scale=tau)
+    proposal_alpha1 = norm(loc=0, scale=tau)
+    proposal_gamma0 = norm(loc=0, scale=tau)
+    proposal_gamma1 = norm(loc=0, scale=tau)
```
Comment (lbotti319) [inline model_code.py:185]: I'm wondering if these should actually be centered at 0. Would it make more sense to center them at the initial values, or some other estimate of the true value of these parameters?

Comment (inespancorbo) [reply model_code.py:185]: Yeah so far I sort of randomly chose the initial values but I think we should choose better start values? To make mu (poisson) and p (bernoulli) reasonable and not super large as our absences count and p don't seem to be extremely large when looking at the observed absences count and higher_yes. We can center them at these better start values or maybe center them at the previous sampled value? Would proposal_alpha0 = norm(loc = alphas0[i-1], tau) make sense?

Comment (lbotti319) [reply model_code.py:185]: > I think we should choose better start values?

What are your thoughts on doing a quick logistic regression? Where the target variable is missingness. The coefficients of the logistic regression can be our start values and our centers.

Comment (lbotti319) [reply model_code.py:185]: I'm more than happy to handle this in a later PR though

Comment (inespancorbo) [reply model_code.py:185]: Pushed the notebook with the logistic. I can let the starting values be those and push again. We still need to think about the other covariate -- assuming normality on G2 does throw nonsensical values. We can go with G2 but maybe one of you wants to work parallel to me and figure out a way of getting positive values.

**Comment by inespancorbo** [comment]
Pushed the notebook with the logistic. I can let the starting values be those and push again. We still need to think about the other covariate -- assuming normality on G2 does throw nonsensical values. We can go with G2 but maybe one of you wants to work parallel to me and figure out a way of getting positive values.

**Comment by inespancorbo** [comment]
> I think this is on the right track. It will be a lot easier to evaluate the quality of the imputation once we have complete results with which to compare it.

Luke has the complete results I think? We can do some sort fo comparison now and see what the quality of the imputation is so far -- I have plots for the posterior distributions for sigma2 and beta in the notebook. Assuming absences is the 2nd covariate with missing values. 

As mentioned above or somewhere else, need to chose starting values for the alphas and gammas, probably different tuning parameters for the proposal distributions of the alphas and gammas, and maybe this gets us to a better place. Can also assume different priors (instead of normal) for the alphas and gammas -- leads to different models and we can do model selection with some of the bayesian metrics Prof Meyer introduced in class. Ideas for if we have time or if anyone wants to think about.

**Comment by lbotti319** [comment]
The MCAR notebook generates complete results. I'll modify it to output to the correct directory

File: src/model_code.py [hunk 4]
```diff
@@ -96,6 +97,202 @@ def running_mean(x):
...
+        log_u = np.log(uniform.rvs())
+        log_r = (
+                dist_log_gamma0(gamma0_star, gamma1, absences_sim[:, i-1], age_missing_absences, loc, scale) -
+                dist_log_gamma0(gamma0, gamma1, absences_sim[:, i-1], age_missing_absences, loc, scale)
+        )
+        if log_u < log_r:
+            gamma0 = gamma0_star
+            accepts_gamma0 += 1
+
+        proposal_gamma1 = norm(loc=gamma1, scale=taus[3])
```
Comment (lbotti319) [inline src/model_code.py:253]: I think we should handle the gammas the same way we do the alphas

Comment (inespancorbo) [reply src/model_code.py]: I pushed with updates

**Comment by inespancorbo** [comment]
Can we approve this PR so at least have the results and code in main? we can handle tweaks in subsequent PRs.

**Review by lbotti319 — APPROVED** [review]

**Pull request merged**